converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Observation/ObservationUpsertHandler.cs

```diff
@@ -0,0 +1,105 @@
+            EnsureArg.IsNotNull(context, nameof(context));
+            EnsureArg.IsNotNull(context.ChangeFeedEntry, nameof(context.ChangeFeedEntry));
+            EnsureArg.IsNotNull(context.Request, nameof(context.Request));
+
+            IResourceId patientId = context.Request.Patient.ResourceId;
+            IResourceId imagingStudyId = context.Request.ImagingStudy.ResourceId;
+            ChangeFeedEntry changeFeedEntry = context.ChangeFeedEntry;
+
+            // Parse all observations out of the dataset
+            ParsedObservation parsedObservations = ObservationParser.CreateObservations(
+                changeFeedEntry.Metadata,
+                patientId.ToResourceReference(),
+                imagingStudyId.ToResourceReference()
+            );
+            Collection<Observation> doseSummaries = parsedObservations.DoseSummaries;
+            Collection<Observation> irradiationEvents = parsedObservations.IrradiationEvents;
+
+            if (!doseSummaries.Any())
+            {
+                return null;
+            }
+
+            // operate on the pre-condition that their _should_ only be one DoseSummary per ImagingStudy.
+            // If multiple observations are found; just use the first one.
+            Observation doseSummary = doseSummaries[0];
+
+            // Filter out existing observations
+            Identifier imagingStudyIdentifier = imagingStudyId.ToResourceReference().Identifier;
+            IEnumerable<Observation> existingDoseSummariesAsync = imagingStudyIdentifier != null
+                ? await _fhirService
+                    .RetrieveObservationsAsync(
+                        imagingStudyId.ToResourceReference().Identifier,
+                        cancellationToken)
+                : new List<Observation>();
+
+            var existingDoseSummaries =
+                existingDoseSummariesAsync.ToList();
+            bool isExisting = existingDoseSummaries.Any();
+
+            FhirTransactionRequestEntry transactionRequestEntry;
+            if (isExisting)
+            {
+                transactionRequestEntry = new FhirTransactionRequestEntry(
+                    FhirTransactionRequestMode.Update,
+                    new Bundle.RequestComponent()
+                    {
+                        Method = Bundle.HTTPVerb.PUT,
+                        Url = $"{ResourceType.Observation.GetLiteral()}/{existingDoseSummaries[0].Id}"
+                    },
+                    existingDoseSummaries[0].ToServerResourceId(),
+                    doseSummary);
+            }
+            else
+            {
+                transactionRequestEntry = new FhirTransactionRequestEntry(
+                    FhirTransactionRequestMode.Create,
+                    new Bundle.RequestComponent()
+                    {
+                        Method = Bundle.HTTPVerb.POST,
+                        Url = ResourceType.Observation.GetLiteral()
+                    },
+                    new ClientResourceId(),
```

Comment:
You are looking for observation with StudyId, but not using it during create of the observation. Also is this the right way to create a FHIR resource with profiles?

Reply:
From my understanding from the spec at <https://confluence.hl7.org/pages/viewpage.action?pageId=97464644#RadiationDoseSummaryforDiagnosticProceduresonFHIR-DoseSummaryObservationProfile>, a _profile_ for a given FHIR resource is just an instance of that resource with a pre-defined `code.coding`. So the thing that makes an _Observation_ resource a _Dose Summary Profile_ is the fact that we set the `code.coding` to `{system: "https://loinc.org", code: "75369-6", display: "Radiation exposure and protection information"}`

Reply:
Generally, you would apply a profile to the FHIR resource by setting the `meta.profile` property to some canonical URI. In this case since it's seems to be a work in progress, I'm not sure there's one to use. The profile in a FHIR sense is a restriction on what is considered valid for a given resource.

An example of this is the [US Core profiles](https://www.hl7.org/fhir/us/core/index.html) and an [example observation](https://www.hl7.org/fhir/us/core/Observation-blood-glucose.json.html) defined as being in conformance with that profile: